### PR TITLE
add VectorTools::Project to do L2 projection

### PR DESCRIPTION
**Pull request by davydden**
this is a generalisation of `projct` to `mpi` with a possiblity to do projection of several fields but doing assembly and preconditioner initialization only once. 

p.s. This could go to a different header.

TODO:

- [x] add the same for pojreciton from `MatrixFree` quadrature data
- [x] test projection from quadrature data
- [x] test projection from a function

@jppelteret @kronbichler ping.

### CODE REVIEW

**Comment by davydden** [comment]
all the projection i wanted are covered now: (i) from a function, (ii) from quadrature with standard cell iterators and (iii) from matrix-free cell groups.

Would appreciate the feedback on interface, which can be bended any direction.

**Comment by davydden** [comment]
For example, an alternative is to implement functions for parallel assembly of the mass matrix and then rework the projection class into separate functions, which would take inverse operator as an input parameter.

**Comment by masterleinad** [comment]
@davydden I guess you have seen #3262? ;-)

**Comment by davydden** [comment]
@masterleinad : I must have seen it, but forgot about it. 

Anyway, the two are quire different actually:

Here:

- Here the focus is on scalar-valued data which can be either `Function` or quadrature data (including matrix-free data). Thus use of std::function everywhere.

There:

- You use standard deal.II Function, which could give a bit advantage if you project vector-valued function on FE field with more than one component with interpretation that each component is a component of this vector.
- You do not support projection from quadrature data, including matrix-free. For that, we would need more `create_right_hand_side()` with `std::function`.
- Your implementation won't work for non-homogeneous constraints, there will be additional contribution to the RHS from assembly. 

But I do like the usage of `MatrixFree` for mass-matrix in your case. This eliminates the need of doing assembly a-priori in a class like what i do here.

----


Bottom line is: I like your approach better, but it lacks functionality of this PR? Can you try enhancing your PR to cover my unit tests?

**Comment by kronbichler** [comment]
@davydden @masterleinad It would indeed be great if we could find a common solution. A question is what to do about the template arguments `n_components` and `fe_degree`. It is a bit nasty that one has to specify them. One could get around them by creating additional functions that do not have those template parameters and refer to the templated functions by pre-compiling common cases (say, degree 1 through 6, components 1 through dim+1) and bailing out otherwise. A user interested in another combination could still call the templated function. (In my private branch I have an `FEEvaluation` object that does not require template arguments on the degree and quadrature points - when that branch gets merged in a few months, we can set `fe_degree=-1` for the non-templated default case and would not even have to bail out any more.)

**Comment by kronbichler** [comment]
As to the `std::function` interface: We could probably add an additional function with this argument?

**Comment by kronbichler** [comment]
One more remark: Should we do anything special for the case of `FE_DGQ`? In that case, one can directly invert the mass matrix (as it is block-diagonal) with the tools provided by `MatrixFreeOperators::CellwiseInverseMassMatrix`. No need to call an iterative solver in that case. I could ask one of my students to contribute that code. (Even though the DG case is only barely covered without the yet to be merged private branch which provides face integrals.)

**Comment by davydden** [comment]
> As to the std::function interface: We could probably add an additional function with this argument?

yes, fully agree. We just need another `project()` and another `compute_right_hand_side()` for those. There will be two different: one for FEValues and one for FEEvaluation.

> A question is what to do about the template arguments n_components and fe_degree

i would not bother for now and first make it work with extra template arguments. We can always add later a few prettified functions. 

> Should we do anything special for the case of FE_DGQ?

One should, but probably in a different PR.

**Comment by davydden** [comment]
one more thing related to `n_components`, I was thinking about projection of scalar-valued quadrature data in MatrixFree, so this guy is always 1.

**Comment by bangerth** [comment]
Regarding `std::function` vs `dealii::Function`: Let's consistently use the deal.II `Function` class. There are adapters that take a `std::function` object and make it look like a `dealii::Function` but not the other way around (because `std::function` does not know about the number of components).

**Comment by davydden** [comment]
> Let's consistently use the deal.II Function class. 

you would need to give it up anyway because to project from quadrature data the `std::function` should take a cell iterator (or an integer in matrix-free case) and a quadrature point number. In other words, dealii::Function won't help here to the best of my knowledge.

**Comment by davydden** [comment]
> std::function does not know about the number of components

sure, but it's a completely different story, I am speaking about projection of scalar-valued quadrature data. If someone needs to project a Tensor, he would need to do it component-wise anyway.

**Comment by davydden** [comment]
closing in favour of https://github.com/dealii/dealii/pull/3290
